kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs

```diff
@@ -0,0 +1,420 @@
+
+        // # Reorder the graph of dependencies and compute the strictly
+        // connex components to:
+        // - compute the order in which to extract the definitions
+        // - find the recursive definitions
+        // - group the mutually recursive definitions
+        let reordered_decls = charon_lib::reorder_decls::compute_reordered_decls(&ccx);
+        ccx.translated.ordered_decls = Some(reordered_decls);
+
+        //
+        // =================
+        // **Micro-passes**:
+        // =================
+        // At this point, the bulk of the translation is done. From now onwards,
+        // we simply apply some micro-passes to make the code cleaner, before
+        // serializing the result.
+
+        // Run the micro-passes that clean up bodies.
+        for pass in charon_lib::transform::ULLBC_PASSES.iter() {
+            pass.transform_ctx(&mut ccx)
+        }
+
+        // # Go from ULLBC to LLBC (Low-Level Borrow Calculus) by reconstructing
+        // the control flow.
+        charon_lib::ullbc_to_llbc::translate_functions(&mut ccx);
+
+        trace!("# LLBC resulting from control-flow reconstruction:\n\n{}\n", ccx);
+
+        // Run the micro-passes that clean up bodies.
+        for pass in charon_lib::transform::LLBC_PASSES.iter() {
+            pass.transform_ctx(&mut ccx)
+        }
+
+        // Print the LLBC if requested. This is useful for expected tests.
+        if queries.args().print_llbc {
+            println!("# Final LLBC before serialization:\n\n{}\n", ccx);
+        } else {
+            debug!("# Final LLBC before serialization:\n\n{}\n", ccx);
+        }
+
+        // Display an error report about the external dependencies, if necessary
+        ccx.errors.report_external_deps_errors();
+
+        let crate_data: charon_lib::export::CrateData = charon_lib::export::CrateData::new(&ccx);
+
+        // No output should be generated if user selected no_codegen.
+        if !tcx.sess.opts.unstable_opts.no_codegen && tcx.sess.opts.output_types.should_codegen() {
+            // # Final step: generate the files.
+            // `crate_data` is set by our callbacks when there is no fatal error.
+            let mut pb = llbc_file.to_path_buf();
+            pb.set_extension("llbc");
+            println!("Writing LLBC file to {}", pb.display());
+            if let Err(()) = crate_data.serialize_to_file(&pb) {
+                tcx.sess.dcx().err("Failed to write LLBC file");
+            }
+        }
+
+        (items, None)
+    }
+}
+
+impl CodegenBackend for LlbcCodegenBackend {
+    fn metadata_loader(&self) -> Box<MetadataLoaderDyn> {
+        Box::new(rustc_codegen_ssa::back::metadata::DefaultMetadataLoader)
+    }
+
+    fn provide(&self, providers: &mut Providers) {
+        provide::provide(providers, &self.queries.lock().unwrap());
+    }
+
+    fn print_version(&self) {
+        println!("Kani-llbc version: {}", env!("CARGO_PKG_VERSION"));
```

Comment:
Are we using `llbc` or `aeneas`? We should try to be coherent and only use one of them, don't you think?

Reply:
I wouldn't use CBMC as a model of consistency. :laughing:

I think this is a usability question, and being consistent and using a meaningful name might be the way to go here. I don't think any of our users really know what `goto` is, and using it in a user visible manner is probably more confusing than anything else. Try googling "goto" and you'll see what I'm saying. LOL

My main question to you is, what feature are the users enabling, `aeneas`, `llbc` or `lean`? Which ones are they going to be interacting with?

Reply:
I'd say it depends on whether you want to use other tools than Aeneas on the output llbc. We have Eurydice that compiles llbc to C and we're developing simple static analyses at the moment.

Reply:
@celinval My thinking was that the kani-compiler is an internal binary that is not meant to be used directly. What the kani-compiler is concerned with is to take MIR and translate it to GOTO or LLBC. The kani-driver is the  binary users interact with, and thus the options at that level should mention Lean or Aeneas.

What do you think?

Reply:
For internal names in the compiler, I think `llbc` or `aeneas` would make sense. For user visible things, I think it will depend on what will be visible for the user. If users are not expected to interact with aeneas directly, I would prefer using Lean.